RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc

```diff
@@ -216,21 +248,200 @@ void CAHitQuadrupletGenerator::hitNtuplets(const IntermediateHitDoublets::Region
                            SeedingLayerSetsHits::LayerIndex outer) {
     return pair.innerLayerIndex() == inner && pair.outerLayerIndex() == outer;
   };
-  fillGraph(layers, g, hitDoublets,
-            [&](const SeedingLayerSetsHits::SeedingLayer& inner,
-                const SeedingLayerSetsHits::SeedingLayer& outer,
-                std::vector<const HitDoublets *>& hitDoublets) {
-      using namespace std::placeholders;
-      auto found = std::find_if(regionLayerPairs.begin(), regionLayerPairs.end(),
-                                std::bind(layerPairEqual, _1, inner.index(), outer.index()));
-      if(found != regionLayerPairs.end()) {
-        hitDoublets.emplace_back(&(found->doublets()));
-        return true;
-      }
-      return false;
-    });
+  edm::ESHandle<TrackerTopology> tTopoHand;
+  es.get<TrackerTopologyRcd>().get(tTopoHand);
+  const TrackerTopology *tTopo=tTopoHand.product();	
+
+  const int numberOfHitsInNtuplet = 4;
+  std::vector<CACell::CAntuplet> foundQuadruplets;
+
+
+
+ int index =0;
+ for(const auto& regionLayerPairs: regionDoublets) {
+
+	 const TrackingRegion& region = regionLayerPairs.region();
+	 hitDoublets.clear();
+	 foundQuadruplets.clear();
+	  if (index == 0){   
+	  	createGraphStructure(layers, g);
+	  }
+	  else{  
+		clearGraphStructure(layers, g);
+	  }
+
+	  fillGraph(layers, g, hitDoublets,
+	            [&](const SeedingLayerSetsHits::SeedingLayer& inner,
+	                const SeedingLayerSetsHits::SeedingLayer& outer,
+	                std::vector<const HitDoublets *>& hitDoublets) {
+	      using namespace std::placeholders;
+	      auto found = std::find_if(regionLayerPairs.begin(), regionLayerPairs.end(),
+	                                std::bind(layerPairEqual, _1, inner.index(), outer.index()));
+	      if(found != regionLayerPairs.end()) {
+	        hitDoublets.emplace_back(&(found->doublets()));
+	        return true;
+	      }
+	      return false;
+	    });
+
+	CellularAutomaton ca(g);
+
+	ca.createAndConnectCells(hitDoublets, region, caThetaCut,
+			caPhiCut, caHardPtCut);
+
+	ca.evolve(numberOfHitsInNtuplet);
+
+	ca.findNtuplets(foundQuadruplets, numberOfHitsInNtuplet);
+
+
+	const QuantityDependsPtEval maxChi2Eval = maxChi2.evaluator(es);
+
+ 	// re-used thoughout, need to be vectors because of RZLine interface
```

Comment:
Maybe we could now drop the ", need to be vectors..." part of the comment, as all of them have been arrays already some time.